src/server/pkg/obj/testing/deployment_test.go

```diff
@@ -0,0 +1,359 @@
+	require.NoError(t, err)
+
+	var port int32
+	for _, servicePort := range pachd.Spec.Ports {
+		if servicePort.Name == "api-grpc-port" {
+			port = servicePort.NodePort
+		}
+	}
+	require.NotEqual(t, 0, port)
+
+	// Get the IP address of the nodes (any _should_ work for the service port)
+	nodes, err := kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
+	require.NoError(t, err)
+
+	// Minikube 'Hostname' address type didn't work when testing, use InternalIP
+	var address string
+	for _, addr := range nodes.Items[0].Status.Addresses {
+		if addr.Type == "InternalIP" {
+			address = addr.Address
+		}
+	}
+	require.NotEqual(t, "", address)
+
+	// Connect to pachd
+	tu.WaitForPachdReady(t, namespace)
+	client, err := client.NewFromAddress(fmt.Sprintf("%s:%d", address, port))
+	require.NoError(t, err)
+	return client
+}
+
+func makeManifest(t *testing.T, backend assets.Backend, secrets map[string][]byte, opts *assets.AssetOpts) string {
+	manifest := &strings.Builder{}
+	jsonEncoder, err := serde.GetEncoder("json", manifest, serde.WithIndent(2), serde.WithOrigName(true))
+	require.NoError(t, err)
+
+	// Create a wrapper encoder that rewrites NodePort ports to be randomly
+	// assigned so that we don't get collisions across namespaces and can run
+	// these tests in parallel.
+	encoder := &ManifestRewriter{Encoder: jsonEncoder}
+
+	// Use a separate hostpath on the kubernetes host for each deployment
+	hostPath := fmt.Sprintf("/var/pachyderm-%s", opts.Namespace)
+	err = assets.WriteAssets(encoder, opts, backend, assets.LocalBackend, 1, hostPath)
+	require.NoError(t, err)
+
+	err = assets.WriteSecret(encoder, secrets, opts)
+	require.NoError(t, err)
+
+	return manifest.String()
+}
+
+func withManifest(t *testing.T, backend assets.Backend, secrets map[string][]byte, callback func(namespace string, pachClient *client.APIClient)) {
+	namespaceName := tu.UniqueString("deployment-test-")
+	opts := &assets.AssetOpts{
+		StorageOpts: assets.StorageOpts{
+			UploadConcurrencyLimit:  assets.DefaultUploadConcurrencyLimit,
+			PutFileConcurrencyLimit: assets.DefaultPutFileConcurrencyLimit,
+		},
+		PachdShards:                16,
+		Version:                    "latest",
+		LogLevel:                   "info",
+		Namespace:                  namespaceName,
+		RequireCriticalServersOnly: assets.DefaultRequireCriticalServersOnly,
+		WorkerServiceAccountName:   assets.DefaultWorkerServiceAccountName,
+		NoDash:                     true,
+		LocalRoles:                 true,
+	}
+
+	manifest := makeManifest(t, backend, secrets, opts)
+
+	kubeClient := tu.GetKubeClient(t)
+	namespace := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespaceName}}
+	_, err := kubeClient.CoreV1().Namespaces().Create(namespace)
+	require.NoError(t, err)
+
+	if cleanup {
+		defer func() {
+			err := kubeClient.CoreV1().Namespaces().Delete(namespaceName, nil)
+			require.NoError(t, err)
+		}()
+	}
+
+	cmd := tu.Cmd("kubectl", "apply", "--namespace", namespaceName, "-f", "-")
```

Comment:
At one point I spent a lot of time trying to figure out how to do this with the Go client and never figured it out.

Reply:
Yeah, I looked briefly and it seemed like you just have to create each type of object manually, so I noped out.